src/util/screensaver.cpp

```diff
@@ -183,100 +209,122 @@ void ScreenSaverHelper::triggerUserActivity()
-{
-    if (!s_sendActivity) {
-        // If the D-Bus method didn't work, let's try the Xlib method.
-        const char* name = ":0.0";
-        Display *display;
-        if (getenv("DISPLAY"))
-            name=getenv("DISPLAY");
-        display=XOpenDisplay(name);
-        XResetScreenSaver(display);
-        XCloseDisplay(display);
-        return;
-    }
-    
-    s_sendActivity = false;
+
+void ScreenSaverHelper::inhibitInternal() {
     if (!QDBusConnection::sessionBus().isConnected()) {
-        qWarning("Cannot connect to the D-Bus session bus.\nTo start it, run:\n"
-                "\teval `dbus-launch --auto-syntax`");
+        qWarning() << "Cannot connect to the D-Bus session bus";
         return;
     }
-    s_sendActivity = false;
-    QString errors;
-    for (int i=0; USERACTIVITY[i][0] != nullptr; i++ ) {
-        QDBusInterface iface(USERACTIVITY[i][0], USERACTIVITY[i][1], USERACTIVITY[i][2], 
-            QDBusConnection::sessionBus());
+
+    // Call uninhibit if the handle / cookie has not been reset
+    if (!s_session_handle.isEmpty() || s_inhibit_cookie > 0) {
+        uninhibit();
+    }
+
+    for (const Inhibitor& inhibitor : kDBusInhibitors) {
+        // Iterate through D-Bus idle inhibitors and select the first available
+        QDBusInterface iface(inhibitor.service,
+                inhibitor.object_path,
+                inhibitor.interface,
+                QDBusConnection::sessionBus());
         if (iface.isValid()) {
-            QDBusReply<void> reply = iface.call(USERACTIVITY[i][3]);
+            QDBusReply<uint32_t> reply = iface.call("Inhibit",
+                    "org.mixxx.Mixxx",
+                    "Mixxx is running");
             if (reply.isValid()) {
-                s_sendActivity = true;
-                break;
+                s_inhibitor = &inhibitor;
+                s_inhibit_cookie = reply.value();
+                s_enabled = true;
+                qDebug() << "Idle inhibitor" << inhibitor.interface << "enabled";
+                return;
             } else {
-                errors = errors +  "\nCall to inhibit for " + USERACTIVITY[i][0] + " failed: " 
-                    + reply.error().message();
+                qWarning() << "Inhibit for"
+                           << inhibitor.interface << "failed: "
+                           << reply.error().message();
             }
-        } 
+        } else {
+            qDebug() << "Interface" << inhibitor.interface << "is not valid";
+        }
     }
-    if (!s_sendActivity) {
-        qWarning() << "Could not send activity using the registered DBus methods. " 
-        << "Errors were: " << errors << 
-        "\nWill try to use the Xlib XResetScreensaver instead.";
+
+    // No standard D-Bus inhibitors found, so check if we're running inside a
+    // Flatpak sandbox and try to use the portal instead
+    if (std::getenv("container") && std::string(std::getenv("container")) == "flatpak") {
```

Comment:
don't 
```suggestion
    char* container = std::getenv("container");
    if (container != nullptr && strcmp(container, "flatpak") == 0) {
```

Reply:
I'd go one step further: Why not encapsulate that into an `isRunningAsFlatpak()` utility method for clarity? Either in this file, or mabe even as `mixxx::EnvironmentHelper::isRunningAsFlatpak()` in a new `src/util/environmenthelper.h` file.

Reply:
Thank you, that is an excellent idea! A simple bool function should be a good fit for this and will make the code easier to read.